Add namespaces filtering in header

goock opened the pull request:
Signed-off-by: Marcin Gucki <mgucki@coreweave.com>

<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing!
 -->

### Description of the change

<!-- Describe the scope of your change - i.e. what the change does. -->
This change refers to #2506 and allows passing namespaces in the header. The name of the header field and its value are defined by 2 environment variable:
- `KUBEAPPS_NAMESPACE_HEADER` - the name of the header field, e.g. `X-Consumer-Groups`
- `KUBEAPPS_NAMESPACE_PATTERN` - defines regular expression with the single matching group for the namespace name, e.g. `namespace:([\w-]+)`

For 2 namespaces, the header may look like this: `X-Consumer-Groups: namespace:ns1, namespace:ns2`.


### Benefits

<!-- What benefits will be realized by the code change? -->
It prevents looping over all namespaces in the cluster.
We have too many namespaces and even with parallel workers it takes too long and the number of namespaces is still growing.

The additional header will be injected by our API gateway. This way Kubeapps will only loop over a few namespaces.


### Possible drawbacks

<!-- Describe any known limitations with your change -->
- It applies only to narrow consumers with big infrastructure like ours.
- The deployment of Kubeapps needs additional customization to set the environment variables.

### Applicable issues

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
  - fixes #2506

### Additional information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here.-->

mecampbellsoup commented on pkg/http-handler/http-handler.go:
```diff
@@ -75,6 +76,34 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
+		return []string{}
+	}
+	r := regexp.MustCompile(headerPattern)
+	headerNamespacesOrigin := strings.Split(req.Header.Get(headerName), ",")
```
Why do you need to split the header name like this out of curiosity? Is "get namespaces from multiple different request headers" supported behavior? I thought `KUBEAPPS_NAMESPACE_HEADER` would be just one string value e.g. `X-Consumer-Permissions`.

goock replied:
We can have multiple namespaces in the header defined by `KUBEAPPS_NAMESPACE_HEADER`.

mecampbellsoup replied:
Ah, my bad, `req.Header.Get(headerName)` returns the value of that header. Sorry.

mecampbellsoup commented on pkg/kube/kube_handler.go:
```diff
@@ -843,6 +844,41 @@ func filterActiveNamespaces(namespaces []corev1.Namespace) []corev1.Namespace {
 }
 
+// GetNamespacesFromList return the list of namespaces that the user has permission to access
+func (a *userHandler) GetNamespacesFromList(headerNamespaces []string) ([]corev1.Namespace, error) {
```
In order to DRY this up could you just redefine `GetNamespaces` to accept an optional param like `namespaces []string`? If present, check the inputted list for those that are permitted; if not, fallback to current/existing behavior of "check all".

goock replied:
Both functions are slightly different but certainly, we could merge them.

absoludity replied:
Isn't this function the same as the existing one to filter namespaces for a user? Not sure why it's being re-defined nor why you need to do this filtering in your case (where the header defines the namespaces that the user can access - why can't you trust that?)

goock replied:
I merged them

mecampbellsoup commented on pkg/http-handler/http-handler.go:
```diff
@@ -75,6 +76,34 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
+	var namespaces = []string{}
+	for _, n := range headerNamespacesOrigin {
+		rns := r.FindStringSubmatch(strings.TrimSpace(n))
+		if len(rns) < 2 {
```
Why are you checking `< 2` like this?

goock replied:
The first position in the slice returns the entire string, matching groups start from position no 1.
However, `FindStringSubmatch` returns `nil` when no match found, which I added to the condition.
Thanks for bringing my attention.

absoludity reviewed (commented):
Thanks @goock . I think this is sufficiently small and general that it'd be good to get it landable and included. I've left a couple of comments below.

andresmgot commented on pkg/http-handler/http-handler_test.go:
```diff
@@ -260,12 +263,39 @@ func TestGetNamespaces(t *testing.T) {
 			getNSFunc := GetNamespaces(&kube.FakeHandler{Namespaces: tc.namespaces, Err: tc.err})
 			req := httptest.NewRequest("GET", "https://foo.bar/backend/v1/namespaces", nil)
 
+			if tc.headerName != "" {
+				os.Setenv("KUBEAPPS_NAMESPACE_HEADER", "X-Consumer-Groups")
+				os.Setenv("KUBEAPPS_NAMESPACE_PATTERN", "namespace:([\\w-]+)")
+				req.Header.Set(tc.headerName, tc.headerValue)
+			}
```
I agree with @absoludity, better to use args in this function rather than env vars to avoid also this kind of setup.

andresmgot commented on pkg/http-handler/http-handler.go:
```diff
@@ -75,6 +76,34 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
+	if headerPattern == "" {
+		return []string{}
+	}
+	r := regexp.MustCompile(headerPattern)
```
`MustCompile` panics if the pattern is not valid, better to use `Compile` and gracefully handle the error (mostly because we cannot be sure the patter is valid).

andresmgot reviewed (commented):
Thanks @goock! I agree with the review of the folks here.

goock commented on pkg/http-handler/http-handler.go:
```diff
@@ -75,6 +76,34 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
+// - KUBEAPPS_NAMESPACE_PATTERN is a regular expression and it matches only single regex group,
+//     e.g. KUBEAPPS_NAMESPACE_PATTERN=namespace:([\w-]+)
+func getHeaderNamespaces(req *http.Request) []string {
+	headerName := os.Getenv("KUBEAPPS_NAMESPACE_HEADER")
```
yeah, this is the place that bothers me, args will be much better for sure.

goock replied:
I changed to kubeops' args/

goock commented on pkg/kube/kube_handler.go:
```diff
@@ -241,6 +249,7 @@ type handler interface {
 type AuthHandler interface {
 	AsUser(token, cluster string) (handler, error)
 	AsSVC(cluster string) (handler, error)
+	GetOptions() kubeOptions
```
I pass the kubeops arguments via handler but I'm not completely happy about the solution this way. 
Is there any better way to have access to the arguments in `GetNamespaces(...)`?

mecampbellsoup replied:
@absoludity probably a good Q for you.

absoludity replied:
I'll take a look first thing next week.

absoludity replied:
From a brief look (and note, this is not thought through, but may be worth thinking through), you may not need to pass the arguments themselves down, if you can instead pass down an implementation of an interface which does what you want. That is:

1. Rather than adding the two fields to `handler.Options` you could instead create an interface `NamespaceFetcher` that has the method you require
2. Add that interface as a field on `handler.Options` and have a default implementation that does the current default behavior
3. Add an implementation which does your specific regex implementation, so that it can be instantiated in the main.go when those fields are provided.
4. Pass the interface down to the handlers and just use it.

Again, don't invest effort in this approach unless you're confident it makes sense and would be worthwhile (I've only thought briefly about it). I'll look more closely next week.

absoludity replied:
> I pass the kubeops arguments via handler but I'm not completely happy about the solution this way.
> Is there any better way to have access to the arguments in `GetNamespaces(...)`?

So no, not without refactoring quite a bit which I don't expect you to do. I think having them in the handler is the simplest way (and unfortunately you need the `GetOptions()` accessor rather than using a public field since the handler itself is private isome places). So +1 for method you chose.

goock commented on pkg/kube/kube_handler.go:
```diff
@@ -844,9 +863,22 @@ func filterActiveNamespaces(namespaces []corev1.Namespace) []corev1.Namespace {
+		if err != nil {
+			return nil, err
+		}
+		return namespaceList, nil
```
Do we want to return empty list here or continue with the regular namespace check?

absoludity replied:
I think the if/else condition here should only set the `namespaceList`:

```
if len(trustedNamespaces) > 0 {
    namespaceList = append(namespaceList, trustedNamespaces...)
} else {
    existing fetch from API as shown below
}

// Filter namespaces that are in terminating state
namespaceList = filterActiveNamespaces(namespaceList)
...
```

Also, it's not clear to me why you want to use `filterAllowedNamespaces` in the case where the namespaces have been provided as input? Can't we assume that the user can access these if we've been told so at this point? (`filterAllowedNamespaces` is only used when we know we've gotten a list that isn't restricted to those for the user only).

mecampbellsoup replied:
> Also, it's not clear to me why you want to use `filterAllowedNamespaces` in the case where the namespaces have been provided as input? Can't we assume that the user can access these if we've been told so at this point? (`filterAllowedNamespaces` is only used when we know we've gotten a list that isn't restricted to those for the user only).

What happens if there is some drift though between the whitelisted namespaces and the ones the user can _actually_ access per k8s RBAC?

antgamdia replied:
Currently, we are completely relying on the namespaces returned by the header. Whereas it is supposed to be an admin configuration, what if the info returned by the API GW (headers) and does not match the one hold by the cluster?
Shouldn't we still perform the filtering based on the permissions?  

If not, what would be the user-facing expected error if that happens to occur?

mecampbellsoup replied:
> Shouldn't we still perform the filtering based on the permissions?

Sorry, to be clear, by 'permissions' do you mean API gateway-specified permissions, or the k8s RBAC permissions?

antgamdia replied:
Sorry, yes, I meant the k8s RBAC permissions. Current implementation is not running `filterAllowedNamespaces` to check whether the user has permission to write secrets. So, in case the API gw gives a wrong namespace, we would be skipping the verification and wouldn't notice the error, no?

mecampbellsoup replied:
I believe that's correct and in that case we'd need to now handle k8s API auth errors from the actually-not-whitelisted whitelisted namespaces... cc @goock

absoludity replied:
> Sorry, yes, I meant the k8s RBAC permissions. Current implementation is not running `filterAllowedNamespaces` to check whether the user has permission to write secrets. So, in case the API gw gives a wrong namespace, we would be skipping the verification and wouldn't notice the error, no?

I had thought the intention was that if the headers are set, they are used as is and that it is an issue with the configuration if the gateway is including namespaces for which it should not (ie. no security issue, just the user *won't* be able to use them as RBAC will not allow it)? We could ensure that `filterAllowedNamespaces()` is also called for the defined namespaces, but can end up back in a similar position if many namespaces are configured.

If we assume that any configured namespaces from the header are correct, and they are *incorrectly* configured (ie. users able to select namespaces to which they cannot write), the error would be noticed when users select the namespace (if they don't have RBAC to read) or deploy to that namespace (if they can read, but can't write). On the other hand, if we do filter the namespaces provided by the header, it will hide the configuration error (users just won't have the namespace to select?)

IMO, the latter would be more of an issue (not detecting the configuration error where the gateway is adding namespaces for a user that it should not because we automatically filter them out), but don't feel strongly about it.

antgamdia replied:
>  We could ensure that filterAllowedNamespaces() is also called for the defined namespaces, but can end up back in a similar position if many namespaces are configured. 
> (...)  if we do filter the namespaces provided by the header, it will hide the configuration error (users just won't have the namespace to select?)

Mmm, I see. Yes, it makes sense just avoiding double checks if it is the gw which is performing this logic. As you said, if we filter them out, we will be hiding the configuration error, which is way worse than just showing an RBAC error.

Ok then, I just wanted to check it with you. Thanks for your comments; go ahead!

mecampbellsoup commented on chart/kubeapps/templates/kubeops-deployment.yaml:
```diff
@@ -63,6 +63,12 @@ spec:
             {{- if .Values.kubeops.QPS }}
             - --qps={{ .Values.kubeops.QPS }}
             {{- end }}
+            {{- if .Values.kubeops.ns-header-name }}
```
You didn't test this did you? To shame!! 😆 

```
{{ index .Values.kubeops "ns-header-name" }}
```

https://stackoverflow.com/questions/63853679/helm-templating-doesnt-let-me-use-dash-in-names

mecampbellsoup commented on pkg/http-handler/http-handler_test.go:
```diff
@@ -260,6 +262,27 @@ func TestGetNamespaces(t *testing.T) {
+			headerValue:  "nspaces:ns1, nspaces:ns2",
+		},
+		{
+			name:         "it should return the list of namespaces and a 200 when when header does not match KUBEAPPS_NAMESPACE_PATTERN",
```
You need to update these descriptions right? `KUBEAPPS_NAMESPACE_PATTERN` should instead refer to the Helm chart value?

mecampbellsoup commented on pkg/kube/kube_handler_test.go:
```diff
@@ -909,6 +910,45 @@ func TestGetNamespaces(t *testing.T) {
 			expectedNamespaces: []string{"bar"},
 			allowed:            true,
 		},
+		{
```
Could you add a test where the regex value is `''`?

goock replied:
I think you are looking for those 2 tests: https://github.com/kubeapps/kubeapps/pull/2671/files#diff-9a009d80344fc69fac8f2a1d383b6fa3545305bb48ac693d1b894f94ed37853aR305

absoludity replied:
Right, sorry - these ones here are more integration tests for the GetNamespaces handler rather than unit tests for the underlying `GetNamespaces` function. Thanks.

absoludity commented on pkg/kube/kube_handler_test.go:
```diff
@@ -909,6 +910,45 @@ func TestGetNamespaces(t *testing.T) {
+			}},
+			expectedNamespaces: []string{"foo"},
+			allowed:            true,
+		},
```
Similar to the other tests, I'm not sure these tests are testing what you think they're testing. For starters, in your use case, the user client should not be even contacted, as the namespaces are passed in? I don't think you should even be setting `existingNamespaces` for your tests, given that we expect the API server not to be called for the list (especially if you remove the use of `filterAllowedNamespaces` in your code which may cause an indirect query).

I think I'd just have three tests here:

* it returns namespaces from header when specified
* it filters namespaces from header to return only active namespaces
* it returns namespaces from header, despite user and service account forbidden

and for all three, `existingNamespaces` need not be set (as it should be empty).

goock replied:
I review the code a few times and if I'm not wrong, the test does the job. I reduced the test to only 2 cases:
- it returns namespaces from header when specified,
- it returns existing namespaces when `whitelistedNamespaces` is empty i.e. header does not match, it's empty, or one of the kubeops arg is empty.

mecampbellsoup commented on pkg/kube/kube_handler.go:
```diff
@@ -229,7 +237,7 @@ type handler interface {
 	RefreshAppRepository(repoName string, requestNamespace string) (*v1alpha1.AppRepository, error)
 	DeleteAppRepository(name, namespace string) error
-	GetNamespaces() ([]corev1.Namespace, error)
+	GetNamespaces(trustedNamespaces []corev1.Namespace) ([]corev1.Namespace, error)
```
I like something like `presetNamespaces` or `whitelistedNamespaces`, I think that is the intention we want to convey.

goock replied:
I like `whitelistedNamespaces`, this is exactly what we are doing.

antgamdia replied:
Thanks for all the changes! I personally do not think `whitelistedNamespaces` is a good name as it is against the [Inclusive Naming Initiative](https://inclusivenaming.org/language/word-list/) (the Linux Foundation, VMware and other organizations are supporting it). 
Therefore, I'd strongly suggest using an alternative name rather than this one: `allowListedNamespaced` or `allowedNamespaces` perhaps? See what you think @absoludity.

absoludity replied:
Yep, +1 for inclusive naming. Naming is hard :) I think the same intention, in this specific case, is conveyed with `preapprovedNamespaces` or `precheckedNamespaces` ?

goock commented on pkg/http-handler/http-handler.go:
```diff
@@ -75,6 +77,35 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
+	}
+	r, err := regexp.Compile(headerPattern)
+	if err != nil {
+		log.Errorf("unable to compile regular expression: %v", err)
```
Agree 100%. It's gonna be easier to find the bug if there is an error in regex. User, if, wants to use `namespace-header-name` or `namespace-header-pattern` needs to be aware of possible errors in regex. I send the message to log but also pass down the error to the client.

absoludity commented on pkg/http-handler/http-handler_test.go:
```diff
@@ -245,50 +246,93 @@ func TestDeleteAppRepository(t *testing.T) {
-			expectedCode: 200,
-			headerName:   "Y-Consumer-Groups",
-			headerValue:  "namespace:ns1, namespace:ns2",
+			name:               "it should return the existing list of namespaces and a 200 when when header does not match kubeops arg namespace-header-pattern",
```
```suggestion
			name:               "it should return the existing list of namespaces and a 200 when header does not match kubeops arg namespace-header-pattern",
```

absoludity commented on pkg/http-handler/http-handler_test.go:
```diff
@@ -245,50 +246,93 @@ func TestDeleteAppRepository(t *testing.T) {
+			namespaceHeaderOptions: kube.KubeOptions{
+				NamespaceHeaderName:    "X-Consumer-Groups",
+				NamespaceHeaderPattern: "",
+			},
```
Perhaps another case: when the header name matches and the header pattern isn't empty but doesn't result in any matches, what happens? (from your code, looks like you only do the early return with the namespaces from header if there are more than 0, so we'd again expect the existingNamespaces in that case)

goock replied:
@absoludity Do you think about this test case: https://github.com/kubeapps/kubeapps/pull/2671/files#diff-9a009d80344fc69fac8f2a1d383b6fa3545305bb48ac693d1b894f94ed37853aR282
I added an additional case when some of the header namespaces match the regex : https://github.com/kubeapps/kubeapps/pull/2671/files#diff-9a009d80344fc69fac8f2a1d383b6fa3545305bb48ac693d1b894f94ed37853aR326
The test case shows we need to enclose regex in `^...$`.

absoludity replied:
> @absoludity Do you think about this test case: https://github.com/kubeapps/kubeapps/pull/2671/files#diff-9a009d80344fc69fac8f2a1d383b6fa3545305bb48ac693d1b894f94ed37853aR282

Thanks, perfect (not sure if I missed that as the test name says it's based on the header (itself) not matching, whereas it's actually the case I was looking for, where the regex *value* doesn't match).

> I added an additional case when some of the header namespaces match the regex : https://github.com/kubeapps/kubeapps/pull/2671/files#diff-9a009d80344fc69fac8f2a1d383b6fa3545305bb48ac693d1b894f94ed37853aR326

Great, thanks.

> The test case shows we need to enclose regex in `^...$`.

Excellent, +1 for testing :) (though that's a fix for your actual regex matching use-case, no the code itself, but great to demo).

absoludity commented on pkg/http-handler/http-handler_test.go:
```diff
@@ -260,6 +262,27 @@ func TestGetNamespaces(t *testing.T) {
+			expectedCode: 200,
+			headerName:   "X-Consumer-Groups",
+			headerValue:  "namespace:ns1, namespace:ns2",
+		},
```
Thanks for the update.

absoludity commented on pkg/kube/fake.go:
```diff
@@ -88,7 +94,10 @@ func (c *FakeHandler) GetAppRepository(name, namespace string) (*v1alpha1.AppRep
 
 // GetNamespaces fake
-func (c *FakeHandler) GetNamespaces() ([]corev1.Namespace, error) {
+func (c *FakeHandler) GetNamespaces(whitelistedNamespaces []corev1.Namespace) ([]corev1.Namespace, error) {
```
We just need an inclusive name as mentioned.

absoludity commented on pkg/kube/kube_handler.go:
```diff
@@ -844,37 +863,43 @@ func filterActiveNamespaces(namespaces []corev1.Namespace) []corev1.Namespace {
 	namespaceList = filterActiveNamespaces(namespaceList)
-	if err != nil {
-		return nil, err
-	}
```
Woops, thanks.

absoludity reviewed (commented):
Excellent, thanks for the changes and tests. Just keen for a new name, such as `precheckedNamespaces` or `preapprovedNamespaces` instead of `whitelistedNamespaces` and we're good to go, IMO.

absoludity reviewed (approved):
+1 from me. Thanks for working with us to get this in! I'll just wait for a +1 from @antgamdia as well before we land.

mecampbellsoup commented on pkg/http-handler/http-handler.go:
```diff
@@ -75,6 +77,35 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
+		log.Errorf("unable to compile regular expression: %v", err)
+		return namespaces, err
+	}
+	headerNamespacesOrigin := strings.Split(req.Header.Get(headerName), ",")
```
@goock what happens if this returns an empty list?

goock replied:
> @goock what happens if this returns an empty list?

nothing actually, when the slice is empty, the loop `for` just skips, and the entire function `getHeaderNamespaces` returns an empty list of the namespaces.

mecampbellsoup commented:
Right so upstream of that do we fallback to the original behavior where all
namespaces are checked?

On Thu, Apr 22, 2021 at 9:30 AM Marcin Gucki ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/http-handler/http-handler.go
> <https://github.com/kubeapps/kubeapps/pull/2671#discussion_r618401897>:
>
> > @@ -75,6 +77,35 @@ func getNamespaceAndCluster(req *http.Request) (string, string) {
>  	return requestNamespace, requestCluster
>  }
>
> +// getHeaderNamespaces returns a list of namespaces from the header request
> +// The name and the value of the header field is specified by 2 variables:
> +// - headerName is a name of the expected header field, e.g. X-Consumer-Groups
> +// - headerPattern is a regular expression and it matches only single regex group, e.g. ^namespace:([\w-]+)$
> +func getHeaderNamespaces(req *http.Request, headerName, headerPattern string) ([]corev1.Namespace, error) {
> +	var namespaces = []corev1.Namespace{}
> +	if headerName == "" || headerPattern == "" {
> +		return []corev1.Namespace{}, nil
> +	}
> +	r, err := regexp.Compile(headerPattern)
> +	if err != nil {
> +		log.Errorf("unable to compile regular expression: %v", err)
> +		return namespaces, err
> +	}
> +	headerNamespacesOrigin := strings.Split(req.Header.Get(headerName), ",")
>
> @goock <https://github.com/goock> what happens if this returns an empty
> list?
>
> nothing actually, when the slice is empty, the loop for just skips, and
> the entire function getHeaderNamespaces returns an empty list of the
> namespaces.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeapps/kubeapps/pull/2671#discussion_r618401897>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPS7LO7BZCNOIZWIKYP2F3TKAQGFANCNFSM42Z6FRTQ>
> .
>

goock commented:
> Right so upstream of that do we fallback to the original behavior where all namespaces are checked?
> […](#)
> On Thu, Apr 22, 2021 at 9:30 AM Marcin Gucki ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In pkg/http-handler/http-handler.go <[#2671 (comment)](https://github.com/kubeapps/kubeapps/pull/2671#discussion_r618401897)>: > @@ -75,6 +77,35 @@ func getNamespaceAndCluster(req *http.Request) (string, string) { return requestNamespace, requestCluster } +// getHeaderNamespaces returns a list of namespaces from the header request +// The name and the value of the header field is specified by 2 variables: +// - headerName is a name of the expected header field, e.g. X-Consumer-Groups +// - headerPattern is a regular expression and it matches only single regex group, e.g. ^namespace:([\w-]+)$ +func getHeaderNamespaces(req *http.Request, headerName, headerPattern string) ([]corev1.Namespace, error) { + var namespaces = []corev1.Namespace{} + if headerName == "" || headerPattern == "" { + return []corev1.Namespace{}, nil + } + r, err := regexp.Compile(headerPattern) + if err != nil { + log.Errorf("unable to compile regular expression: %v", err) + return namespaces, err + } + headerNamespacesOrigin := strings.Split(req.Header.Get(headerName), ",") @goock <[@goock](https://github.com/goock)> what happens if this returns an empty list? nothing actually, when the slice is empty, the loop for just skips, and the entire function getHeaderNamespaces returns an empty list of the namespaces. — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#2671 (comment)](https://github.com/kubeapps/kubeapps/pull/2671#discussion_r618401897)>, or unsubscribe <[github.com/notifications/unsubscribe-auth/AAPS7LO7BZCNOIZWIKYP2F3TKAQGFANCNFSM42Z6FRTQ](https://github.com/notifications/unsubscribe-auth/AAPS7LO7BZCNOIZWIKYP2F3TKAQGFANCNFSM42Z6FRTQ)> .

if `precheckedNamespaces` in https://github.com/kubeapps/kubeapps/pull/2671/files#diff-a7d0ac1ab1c1fe440361def3cc5c037d41a25511e6c1e7b35084358b6c4f858aR294 is empty, we fallback to the original behaviour.

goock commented:
@absoludity I found leftover from the previous commit:
```
		// if any, use only overrided namespaces
		if len(namespaceList) > 0 {
			return namespaceList, nil
```
I rebased and removed it after this line https://github.com/kubeapps/kubeapps/pull/2671/files#diff-680a1eb150fd209972a67711ba74bb11159483a38c054070ca2d01d8dc919007R870.

antgamdia reviewed (approved):
LGTM! Thank you all for the effort, it's always great to receive contributions. I hope we can release asap :) 

I only have a comment regarding, https://github.com/kubeapps/kubeapps/pull/2671#discussion_r616658841

goock commented:
Thank you guys for the help and comments to make this PR happened.

absoludity commented:
Thanks for your patience getting it generalised!

mecampbellsoup commented:
Needs this fix merged to actually work 😭 

https://github.com/kubeapps/kubeapps/pull/2981
